packages/lodestar/src/sync/subnet/attestation-collector.ts

```diff
@@ -0,0 +1,71 @@
+import {IBeaconChain} from "../../chain";
+import {OpPool} from "../../opPool";
+import {IBeaconConfig} from "@chainsafe/eth2.0-config";
+import {Attestation, CommitteeIndex, Slot} from "@chainsafe/eth2.0-types";
+import {IService} from "../../node";
+import {INetwork} from "../../network";
+import {getCommitteeIndexSubnet} from "../../network/gossip/utils";
+
+export interface IAttestationCollectorModules {
+  chain: IBeaconChain;
+  network: INetwork;
+  opPool: OpPool;
+}
+
+export class AttestationCollector implements IService {
+
+  private readonly config: IBeaconConfig;
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly opPool: OpPool;
+
+  private aggregationDuties: Map<Slot, Set<CommitteeIndex>> = new Map();
+
+  public constructor(config: IBeaconConfig, modules: IAttestationCollectorModules) {
+    this.config = config;
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.opPool = modules.opPool;
+  }
+
+  public async start(): Promise<void> {
+    this.chain.clock.onNewSlot(this.checkDuties);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  public async stop(): Promise<void> {
+  }
+
+  public subscribeToCommitteeAttestations(slot: Slot, committeeIndex: CommitteeIndex): void {
+    this.network.gossip.subscribeToAttestationSubnet(getCommitteeIndexSubnet(committeeIndex));
+    if(this.aggregationDuties.has(slot)) {
+      this.aggregationDuties.get(slot).add(committeeIndex);
+    } else {
+      this.aggregationDuties.set(slot, new Set([committeeIndex]));
+    }
+  }
+
+  private checkDuties = (slot: Slot): void => {
+    const committees = this.aggregationDuties.get(slot) || new Set();
+    committees.forEach((committeeIndex) => {
+      this.network.gossip.subscribeToAttestationSubnet(
+        getCommitteeIndexSubnet(committeeIndex),
+        this.handleCommitteeAttestation
+      );
+      setTimeout(
+        this.unsubscribeSubnet,
```

Comment:
I'm thinking about a race condition when we may subscribe/unsubscribe the same subnet at the same time.

For example, at slot 2 we subscribe at subnet 10 then at the end of slot 2 we unsubscribe by `setTimeout`, but onNewSlot of slot 3 we want to subscribe at subnet 10 again. How can we make sure the earlier unsubscribe happens before this subscribe as it may come at the same time?

Reply:
Nope, this won't happen. I've changed logic of unsubscription. It will only unsubscribe from topic if there is 0 listeners.